lfs/batcher.go

```diff
@@ -51,16 +55,22 @@ func (b *Batcher) acceptInput() {
 	exit := false
 
 	for {
-		batch := b.newBatch()
+		batch := make([]Transferable, 0, b.batchSize)
 	Loop:
 		for len(batch) < b.batchSize {
-			t, ok := <-b.input
-			if !ok {
-				exit = true // input channel was closed by Exit()
-				break Loop
+			select {
+			case t, ok := <-b.input:
+				if ok {
+					batch = append(batch, t)
+				} else {
+					exit = true // input channel was closed by Exit()
+					break Loop
+				}
+			case <-time.After(time.Millisecond * 300):
```

Comment:
I was imagining that original code that was adding the Transferables would call Flush() after it had finished adding, which would guarantee the batcher would send to batchReady at that point regardless of numbers. I agree that you'd have to separate the failures into a different queue (different batcher?) to submit at the end to be truly deterministic & ensure that every batch was the same regardless of timing. It does add a little code complexity but I think knowing that if you submit the same items in the same order that they will always result in the same requests at the end, regardless of timings, is a nice position to be in from a diagnostic point of view.